### PR TITLE
IMPRO-1494: Remove cell methods added by iris collapsed

### PR DESCRIPTION
**Pull request by neilCrosswaite**
Added a utility which calls `cube.collapsed` but returns it without the new `cell_method`.

Applied this function to add the uses of `cube.collapsed`

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver_tests/utilities/cube_manipulation/test_collapsed.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4818437&bid=16021375)

File: improver/cli/extract.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def process(cube: cli.inputcube,
             *,
             constraints: parameters.multi(min=1),
             units: cli.comma_separated_list = None,
-            ignore_failure=False):
+            ignore_failure = False):
```
Comment (neilCrosswaite) [inline improver/cli/extract.py]:
```suggestion
            ignore_failure=False):
```

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver_tests/utilities/cube_manipulation/test_collapsed.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4818437&bid=16021375)

File: improver/utilities/cube_manipulation.py [hunk 2]
```diff
@@ -43,6 +42,25 @@
...
 
 
+def collapsed(cube, *args, **kwargs):
+    """Collapses the cube with given arguments but doesn't keep the cell method
+
+    Args:
+        cube (iris.cube.Cube):
+            A Cube to be collapsed.
+        *args:
+        **kwargs:
```
Comment (bayliffe) [inline improver/utilities/cube_manipulation.py]: Do these need to be included in the doc-string, given that they can be anything?

File: improver/utilities/cube_manipulation.py [hunk 3]
```diff
@@ -43,6 +42,25 @@
...
+
+    Returns:
+        (iris.cube.Cube)
```
Comment (bayliffe) [inline improver/utilities/cube_manipulation.py]: No brackets, followed by colon.

File: improver/ensemble_calibration/ensemble_calibration.py [hunk 4]
```diff
@@ -47,7 +47,8 @@
     check_predictor_of_mean_flag, convert_cube_data_to_2d,
     flatten_ignoring_masked_data)
 from improver.utilities.cube_checker import time_coords_match
-from improver.utilities.cube_manipulation import enforce_coordinate_ordering
+from improver.utilities.cube_manipulation import (enforce_coordinate_ordering,
+    collapsed)
```
Comment (bayliffe) [inline improver/ensemble_calibration/ensemble_calibration.py]: Under indented.

File: improver/nbhood/use_nbhood.py [hunk 5]
```diff
@@ -402,8 +403,10 @@ def process(self, cube):
         # Loop over any extra dimensions
         cubelist = iris.cube.CubeList([])
         for slice_3d in cube.slices([self.coord_masked, yname, xname]):
-            collapsed_slice = slice_3d.collapsed(
-                self.coord_masked, iris.analysis.MEAN, weights=weights)
+
```
Comment (bayliffe) [inline improver/nbhood/use_nbhood.py]: Such a cold lonely void. Why must I be so far from my parent for loop? Please reunite us.

File: improver/nowcasting/optical_flow.py [hunk 6]
```diff
@@ -88,7 +89,8 @@ def _calculate_time_average(wind_cubes, time_coord):
         """Average input cubelist over time"""
         cube = wind_cubes.merge_cube()
         try:
-            mean = cube.collapsed("time", iris.analysis.MEAN)
+
```
Comment (bayliffe) [inline improver/nowcasting/optical_flow.py]: Once again you have placed a rift between us. Why oh why?

File: improver/percentile.py [hunk 7]
```diff
@@ -124,10 +125,13 @@ def process(self, cube):
         # Rename the percentile coordinate to "percentile" and also
         # makes sure that the associated unit is %.
         if n_valid_coords == n_collapse_coords:
-            result = cube.collapsed(
-                self.collapse_coord, iris.analysis.PERCENTILE,
+
```
Comment (bayliffe) [inline improver/percentile.py]: You get the idea....

File: improver_tests/utilities/cube_manipulation/test_collapsed.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import unittest
+
+import iris
+import numpy as np
+
+from improver.utilities.cube_manipulation import collapsed
+from ...set_up_test_cubes import set_up_variable_cube
+
+
+class Test_strip_var_names(unittest.TestCase):
```
Comment (bayliffe) [inline improver_tests/utilities/cube_manipulation/test_collapsed.py]: This is not what this should be called.

File: improver/ensemble_calibration/ensemble_calibration.py (hunk 4 above)
Comment (neilCrosswaite) [reply improver/ensemble_calibration/ensemble_calibration.py]: Done

File: improver/nbhood/use_nbhood.py (hunk 5 above)
Comment (neilCrosswaite) [reply improver/nbhood/use_nbhood.py]: Back with family. I'm sorry `code`, please forgive me.

File: improver/nowcasting/optical_flow.py (hunk 6 above)
Comment (neilCrosswaite) [reply improver/nowcasting/optical_flow.py]: Because a regex went wild.

File: improver/percentile.py (hunk 7 above)
Comment (neilCrosswaite) [reply improver/percentile.py]: This one had it coming.

File: improver/utilities/cube_manipulation.py (hunk 2 above)
Comment (neilCrosswaite) [reply improver/utilities/cube_manipulation.py]: Fair, fixed.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver_tests/utilities/cube_manipulation/test_collapsed.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4818437&bid=16021375)

File: improver/nbhood/use_nbhood.py [hunk 9]
```diff
@@ -402,9 +403,10 @@ def process(self, cube):
         # Loop over any extra dimensions
         cubelist = iris.cube.CubeList([])
         for slice_3d in cube.slices([self.coord_masked, yname, xname]):
-            collapsed_slice = slice_3d.collapsed(
-                self.coord_masked, iris.analysis.MEAN, weights=weights)
+            collapsed_slice = collapsed(slice_3d, self.coord_masked,
```
Comment (gavinevans) [inline improver/nbhood/use_nbhood.py]: I notice that the CollapseMaskedNeighbourhoodCoordinate plugin contains a method called `remove_collapsed_coord_refs`. I think that following your creation of a `collapsed` function, the call to the `remove_collapsed_coord_refs` method could be replaced by just `result.remove_coord(self.coord_masked)`, as the cell method operations in `remove_collapsed_coord_refs` are now redundant following your creation of the `collapsed` function.

File: improver/utilities/cube_manipulation.py [hunk 10]
```diff
@@ -43,6 +42,23 @@
 from improver.utilities.cube_checker import check_cube_coordinates
 
 
+def collapsed(cube, *args, **kwargs):
+    """Collapses the cube with given arguments but doesn't keep the cell method
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver/utilities/cube_manipulation.py]: Sounds good, thank you.

File: improver/utilities/cube_manipulation.py [hunk 11]
```diff
@@ -43,6 +42,23 @@
...
+
+    Args:
+        cube (iris.cube.Cube):
+            A Cube to be collapsed.
+
+    Returns:
+        iris.cube.Cube:
+            A collapsed cube without the new cell_method.
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver/utilities/cube_manipulation.py]: Also better, thank you.

**Comment by fionaRust** [comment]
@arh89  and @neilCrosswaite. What we should probably do is consider raising a PR with Iris to fix this behaviour and see they will let us fix the problem there. This can then be a stop gap until we have access to a stack with the a version of Iris that doesn't add these cell methods.

**Comment by arh89** [comment]
> @arh89 and @neilCrosswaite. What we should probably do is consider raising a PR with Iris to fix this behaviour and see they will let us fix the problem there. This can then be a stop gap until we have access to a stack with the a version of Iris that doesn't add these cell methods.

I'm all for fixing things upstream if we can. If we need a stop-gap solution, sure - but if we can do the same thing with fewer lines of code and a less error prone way, then that's probably also worth considering too. 😃

**Comment by arh89** [comment]
Discussed this with @neilCrosswaite - happy to do with what we have (once this has been reviewed by blue team/others - I haven't reviewed this). We'll need a bit more work so that we can make use of the BasePlugin etc.

File: improver/nbhood/use_nbhood.py (hunk 9 above)
Comment (neilCrosswaite) [reply improver/nbhood/use_nbhood.py]: The `remove_collapsed_coord_refs` method can be removed without breaking any tests, just one line of code needs to be moved out of the loop.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver_tests/utilities/cube_manipulation/test_collapsed.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4818437&bid=16021375)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver_tests/utilities/cube_manipulation/test_collapsed.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4818437&bid=16021375)

File: improver/nbhood/use_nbhood.py [hunk 12]
```diff
@@ -402,16 +384,17 @@ def process(self, cube):
...
             cubelist.append(collapsed_slice)
+
         result = cubelist.merge_cube()
         # Promote any scalar coordinates with one point back to dimension
         # coordinates if they were dimensions in the input cube.
         # Take a slice over the coordinate we are collapsing as we do not
         # expect this in the output cube.
         first_slice = next(cube.slices_over([self.coord_masked]))
         result = check_cube_coordinates(first_slice, result)
+        result.remove_coord(self.coord_masked)
```
Comment (gavinevans) [inline improver/nbhood/use_nbhood.py]: The comment on the line below should be above this line.

Comment (neilCrosswaite) [reply improver/nbhood/use_nbhood.py]: Done

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver_tests/utilities/cube_manipulation/test_collapsed.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4818437&bid=16021375)

**Pull request merged**